Adding automatic reformatting from initial commit

MaxBromberg opened the pull request:
Simply committed hierarchy coordinates as a new project in _algorithms_, though without testing as of yet. New to such projects, I'd appreciate some guidance regarding testing and whatever else's needed. An issue with graph entropy remains to be resolved.

[Paper the algorithms are based on][2], [Earlier git with readme][1], [initial discussion][3]


[1]: https://github.com/MaxBromberg/Hierarchy_Coordinates
[2]: https://arxiv.org/abs/1303.2503
[3]: https://groups.google.com/g/networkx-discuss/c/Kltl3aQAWzU/m/SyQhne4FBgAJ

<!--
Please run black to format your code.
See https://networkx.org/documentation/latest/developer/contribute.html for details.
-->

dschult commented:
Wow -- this is a lot of code and docs and tests..   Thank you!
It might take a while to go through in its entirety but I've got a few comments to start.

- the weakly_connected_components_subgraphs function was removed because it is a simple extension of `weakly_connected_components`.  And we didn't feel that the subgraphs were usually needed, and when they were it isn't clear if they should be subgraph views or subgraph views that are copied to make networkx graphs.  So  we think it is better to let the user build that rather than provide all possible options in a convoluted api.  Something like this -- though I don't think you should actually make a list of all of them. Most likely you want to process one at a time to save memory:
```python
[G.subgraph(cc).copy() for cc in nx.weakly_connected_components(G)]
```
- It seems like the functions in `utils.py` should be available as numpy commands. Can we avoid this whole module?  Perhaps these functions could be "private" functions (start with an underscore and not included in `__all__`) within the main module, or simply replaced by their numpy equivalent if there is one.
- The doc_strings in the main `hierarchy_coordinates.py` module should start with an explanation of what hierarchy coordinates are and why someone might want to use them. Then for the functions themselves the doc_strings should start with a single line that explains what it does in very terse lamguage. This is the line the doc-builder sphinx puts near links in the docs to this function. It should describe the function enough that people know whether to click to find out more.  Then the details can be put in the rest of the doc_string. After the first line, there should be a blank line and then paragraph(s) describing what the function does in more detail, the algorithm used, etc. Then parameters, return values and examples, "See Also" links if reasonable and References.  The first line and paragraphs are all I noticed that could use changes.
- Your subpackage and its main module are named the same thing so namespace collisions are possible. But I'd prefer to make this a single module addition rather than a subpackage if possible (a single hierarchy.py file within algorithms and test file in `algorithms/tests`). But if you expect this to expand I'm open to reconsidering.
- You should connect your new files to the documentation by adding appropriate rst files to `/doc/reference/algorithms`. Similarly, adjust `algorithms/__init__.py` to include your functions in the main networkx namespace.  Take a look at what is in both these places to try to mimic it for your files. 

If you have any questions about any of this, reply here and we'll help out. 

I look forward to learning more about hierarchy coordinates. :}  Thanks!

MaxBromberg commented:
Glad hierarchy coordinates can feature in networkx! 

I restructured as you suggested, though in my latest push, it appears I've made a mistake in formatting examples, and all the checks have thus failed. Whoops; what did I do wrong? 
I'll be happy to fix it (an the below issues) with a bit of guidance; also, let me know if the documentation is a touch too terse. Hierarchy coordinates a a bit tricky to explain in concisely...

Notes:
 - I am awaiting a reply from Prof. Solé to resolve the present graph entropy degeneracy, though both definitions _work_, albeit one is far less computationally efficient, though potentially more as intended.
 - There are various TODO/WIP notes, most of which I just think are general best-practice stuff which I don't know the proper conventions for.
 - Weighted networks will transition from considering only the _largest_ (though that's generally the only) DAG to all resultant DAGs after threshold reduction, though I've yet to implement this. This is where the 
 `[G.subgraph(cc).copy() for cc in nx.weakly_connected_components(G)]`
came in, where the list is made because I'm looking for the max of this set. 
- Computational optimization wasn't exactly a priority, so I'm sure it could be improved all around in time...

dschult commented:
You are right -- there are many errors.  You can see the pytest description of those error by looking at the "details" link for one of the Github Action test sessions that has a red "x".  Scroll down to just below the list of all tests and it shows the errors and what the error is.  You can also get this locally on your system by installing pytest and running `pytest --doctest-modules filename.py`

It looks like 1 error is calling a private function (starts with an underscore) from the `nx.` namespace. You'll have to import those functions from the module's namespace specifically.  

Most of the others are spacing issues.  The examples to a very simple character by character comparison of what you put on the line after the python command and what that python command prints. So extra spaces (or not enough) will make it fail.

:)

MaxBromberg commented:
I see! I corrected the example formatting, but I still get this numpy import error. How do I resolve this? (Should I import functions/classes from numpy rather than numpy as np?)

MaxBromberg commented:
I'm afraid I don't quite understand what I should do regarding the numpy imports; I now import only the needed modules within each function, but the general imports in the tests raise an error. (Though I did not get this via `pytest --doctest-modules test_hierarchy_coordinates.py` or `pytest --doctest-modules hierarchy_coordinates.py`)

dschult commented:
The import within a test function would need the `np = pytest.importorskip("numpy")` syntax to know to skip that test.  But I think you will want to skip **all** tests in this module. Right?  Don't all those tests call functions that use numpy? If so, just do the import at the top of the test module using `pytest.importorskip`.  

I'm trying to say that it is not sufficient to check whether the test code itself uses numpy. If it calls a function that imports numpy you will want to skip that test.

dschult commented:
Other comments:
- You should put `algorithms/hierarchy_coordinates.py` into `networkx/conftest.py` in the `needs_numpy` list near line 200. Sorry I didn't remember to say that before.

- it's our "standard" to leave `np.log` etc using the `np` notation. That makes it clearer to know that the `log` function is coming from numpy and easier to find where numpy is used.  So, for your in-the-function imports, can you switch to `import numpy as np` and revert to the `np.` syntax?

- Would it be "good form" to allow input `A` to be an array-like object by starting the functions with something like `A = np.asarray(A)`?  I'm not sure what the usual way to treat this is. Maybe @rossbar knows.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1228 @@
+            for threshold in thresholds
+        ]
+    # removes isolated nodes (0 in & out degree) from binary nodes. (not needed for consolidation)
+    for index in range(len(nx_graphs)):
+        nx_graphs[index].remove_nodes_from(list(nx.isolates(nx_graphs[index])))
```
```suggestion
    for G in nx_graphs:
        G.remove_nodes_from(nx.isolates(G))
```
You should remove any idioms that use `for index in range(len(...)):`. It can/should almost always be done by iterating over the items themselves. Python makes this faster then the indexes and it is more readable.

Also, can you move the comments from the end of the line to the line previous?  I think that is preferred style for python and it will allow the `black` formatter to put many of these commands on one line too.

MaxBromberg replied:
In-code comments were moved to new lines, which was what I understood to be intended. I fixed the superfluous indices; those that remain are iterating over multiple items. This is not to say that the code could not be optimized considerably; however I have not yet heard from [Professor Solé](http://complex.upf.es/~ricard/Main/RicardSole.html) regarding the implementation of the graph entropy, who I understand to be the intended correspondent for the original paper.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+
+    boundary_layer = max_min_layers(dag, max_layer=forward_entropy)
+    non_extremal_nodes = set(dag.nodes() - boundary_layer)
+    entropy = 0
+    for layer_node in boundary_layer:
+        for non_extremal_node in non_extremal_nodes:
+            if forward_entropy:
+                entropy += P[layer_node][non_extremal_node] * np.log(
+                    dag.out_degree(layer_node)
+                )  # nan for 0 outdegree
+            else:
+                entropy += P[non_extremal_node][layer_node] * np.log(
+                    dag.in_degree(layer_node)
+                )
+    entropy /= len(boundary_layer)
```
```substitution
    deg = dag.out_degree if forward_entropy else dag.in_degree
    e = sum(P[u][v] * np.log(deg(u) for u in boundary_layer for v in non_extremal_nodes)
    entropy = e / len(boundary_layer)
```

MaxBromberg replied:
That's better! I accounted for the transposed indices with `P = P if forward_entropy else P.T` and implemented your change.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+
+    if condensed_nx_graph is None:
+        condensed_nx_graph = weight_nodes_by_condensation(nx.condensation(G))
+    non_cyclic_nodes = [
+        node[0]
+        for node in nx.get_node_attributes(condensed_nx_graph, "weight").items()
+        if node[1] == 1
+    ]
+    total_acyclic_node_weight = sum(
+        [
+            weight
+            for weight in nx.get_node_attributes(condensed_nx_graph, "weight").values()
+        ]
+    )
+    return len(non_cyclic_nodes) / total_acyclic_node_weight
```
```suggestion
    weights = nx.get_node_attributes(condensed_nx_graph, "weight").values()
    total_acyclic_node_weight = sum(weights)
    number_non_cyclic_nodes = sum(1 for wt in weights if wt == 1)
    return number_non_cyclic_nodes / total_acyclic_node_weight
```

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+        )
+    # unweighted (non-binary) check
+    if not np.array_equal(np.unique(nx.to_numpy_array(G)), [0, 1]):
+        o = 0
+        condensed_graphs, original_graphs = node_weighted_condense(
+            nx.to_numpy_array(G),
+            num_thresholds=num_thresholds,
+            threshold_distribution=threshold_distribution,
+        )
+
+        for index in range(len(condensed_graphs)):
+            o += orderability(original_graphs[index], condensed_graphs[index])
+        return o / len(condensed_graphs)
```
```suggestion
        condensed_graphs, original_graphs = node_weighted_condense(
            nx.to_numpy_array(G),
            num_thresholds,
            threshold_distribution,
        )

        o = sum(orderability(OG, CG) for CG, OG in zip(condensed_graphs, original_graphs))
        return o / len(condensed_graphs)
```

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+    """
+    if not G.is_directed():
+        raise nx.NetworkXError("G must be a DiGraph for min/max layer evaluation")
+    if max_layer:
+        return [node for node in G.nodes() if G.in_degree(node) == 0]
+    else:
+        return [node for node in G.nodes() if G.out_degree(node) == 0]
```
```suggestion
    if max_layer:
        return [node for node, deg in G.in_degree if deg == 0]
    else:
        return [node for node, deg in G.out_degree if deg == 0]
```

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+            "G must be a directed acyclic graph for node weighted condensation"
+        )
+
+    node_weights = [
+        len(w) for w in nx.get_node_attributes(condensed_graph, "members").values()
+    ]
+    for node_index in range(len(node_weights)):
+        condensed_graph.nodes[node_index]["weight"] = node_weights[node_index]
```
```suggestion
    for node, attrs in condensed_graph.nodes.data():
        attrs["weight"] = len(attrs["members"])
```

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+                        4,
+                    )
+                )  # linear distribution
+            except:
```
What kind of exception might occur here? Can we provide the exception type, i.e. `except ValueError:`?

MaxBromberg replied:
This was rendered obsolete by the earlier `np.isclose(np.max(A) - np.min(A), 0, 1e-15):` check, so I removed it.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+                )
+            )
+        )
+        members = nx.get_node_attributes(largest_condensed_graphs[-1], "members")
+        node_weights = [len(w) for w in members.values()]
+        for node_index in range(len(node_weights)):
+            largest_condensed_graphs[-1].nodes[node_index]["weight"] = node_weights[
+                node_index
+            ]
```
```suggestion
        for node, attrs in largest_condensed_graphs[-1].nodes.data():
            attrs["weight"] = len(attrs["members"])
```

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+    for condensed_graph in condensed_graphs:
+        largest_condensed_graphs.append(
+            nx.convert_node_labels_to_integers(
+                max(
+                    [
+                        condensed_graph.subgraph(cc).copy()
+                        for cc in nx.weakly_connected_components(condensed_graph)
+                    ],
+                    key=len,
+                )
```
Can the `subgraph(cc).copy()` be pulled out of the `max`? I think `max(nx.weakly_...), key=len)` will give the largest `cc`.

MaxBromberg replied:
Just comparing the two, it appears `max(nx.weakly_...), key=len) ` yields only the set of nodes in the maxima. (Class `Sized`)
```
print(max([condensed_graph.subgraph(cc).copy() for cc in nx.weakly_connected_components(condensed_graph)], key=len))
print(max(nx.weakly_connected_components(condensed_graph), key=len))
```
So in short, it's likely possible, though I know not how.

dschult replied:
```python
max_cc = max(nx.weakly_connected_components(condensed_graph), key=len))
condensed_graph.subgraph(max_cc).copy()
```
I guess it could all be on one line too if you prefer.

MaxBromberg commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+        # raise Exception("Unconnected graph; trivially 0 hierarchy")  # TODO: Raise exception if undefined instead
+        return 0, 0, 0
+
+    o, f, t = 0, 0, 0
+    condensed_graphs, original_graphs = node_weighted_condense(
+        A=np_A,
+        num_thresholds=num_thresholds,
+        threshold_distribution=threshold_distribution,
+    )
+    for index in range(len(condensed_graphs)):
+        t += treeness(condensed_graphs[index])
+        f += feedforwardness(condensed_graphs[index])
+        o += orderability(original_graphs[index], condensed_graphs[index])
+    t /= len(condensed_graphs)
+    f /= len(condensed_graphs)
+    o /= len(condensed_graphs)
+    return t, f, o
```
I left out the list comprehension here so that I could feed orderability the original graphs directly, as I think that's often faster than re-deriving the original (unweighted) graphs, seen in:
`o += orderability(original_graphs[index], condensed_graphs[index])`

MaxBromberg replied:
I left out the list comprehension here so that I could feed orderability the original graphs directly, as I think that's often faster than re-deriving the original (unweighted) graphs, seen in:
`o += orderability(original_graphs[index], condensed_graphs[index])`

dschult replied:
Thanks -- I missed that...   My suggestion should read:

```python
o = sum(orderability(OG, CG) for OG, CG in zip(original_graphs, condensed_graphs)) / len(condensed_graph)
```

MaxBromberg commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1220 @@
+    entropy = 0
+    for start_node in start_layer:
+        for end_node in end_layer:
+            if forward_entropy:
+                paths = list(
+                    nx.all_simple_paths(dag, source=start_node, target=end_node)
+                )
+            else:
+                paths = list(
+                    nx.all_simple_paths(
+                        dag.reverse(copy=True), source=start_node, target=end_node
+                    )
+                )
+            for path in paths:
+                if forward_entropy:
```
yes! Good call, I'm hope this whole algorithm shouldn't be needed, as the analytic version simply called `graph_entropy` would be far faster, and is what is described in their appendix. However, as it disagrees with the apparent method seen in their infographic, I implemented the infographic version as well in hopes of easing explanation for author review. (To agree perfectly with the infographic, change np.log to np.log2, as I have done in latest commit)

MaxBromberg commented:
Thanks for the feedback; I implemented what I could, modified if needed. I abstained from list (sum) comprehension in `average_hierarchy_coordinates.py`, (see comment) but otherwise revised as suggested. 
...But apparently I did something that led to all the checks failing? I tested beforehand a lá `pytest --doctest-modules ...`, so not sure what I did :/

dschult commented:
The test failures are definitely not something you did. It looks like a widespread issue -- probably due to a new version of one of the packages used on our testing system  maybe `decorator`.  We'll look into that. If your branch passes tests locally you are probably good. We'll just rerun the tests once the other problem is fixed.

I think the `np.log` vs `np.log2` issue is a miscommunication about their infographic.  But it's hard to know for sure. I'm not bothered by using either but it would be good to track it down as you say.  I'll take a closer look at all this stuff soon. :}

dschult commented:
I pushed the recent changes from `main` onto this branch. That includes the fix needed for tests to work correctly.
You will need to pull them to your local computer from your github repository using something like 
`git pull origin adding-hierarchy-coords`.  
Hopefully that is smooth.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1166 @@
+    L_GC = len(recursive_leaf_removal(DAG))
+
+    if forward_entropy:
+        # as w sklearn: B_prime = normalize(nx.to_numpy_array(dag), axis=1, norm='max')  # Row normalization
+        B_prime = _matrix_normalize(nx.to_numpy_array(dag), row_normalize=True)
```
It looks like the comments normalizes using the max, but this code normalizes the sum.
Also, you might consider replacing `row_normalize=True` with `axis=1` which matches numpy

dschult commented on networkx/algorithms/tests/test_hierarchy_coordinates.py:
```diff
@@ -0,0 +1,225 @@
+    G = nx.DiGraph(center_cycle)  # unit weights by default
+    pytest.raises(
+        nx.NetworkXError, nx.weight_nodes_by_condensation, G
+    )  # Must be condensed (acyclic)
```
Can you move this comment to the line prior to the pytest?
And other comments in this file that end a line...   Thanks!

dschult commented on networkx/algorithms/tests/test_hierarchy_coordinates.py:
```diff
@@ -0,0 +1,225 @@
+    assert len(condensed_Gs) == len(
+        binary_Gs
+    )  # Not necessarily equal to num thresholds, as empty graphs are dropped
+    for index in range(len(binary_Gs)):
+        assert np.array_equal(
+            np.unique(nx.to_numpy_array(binary_Gs[index])), [0, 1]
+        )  # binary checks
+        assert np.array_equal(np.unique(nx.to_numpy_array(condensed_Gs[index])), [0, 1])
```
```suggestion
    for binaryG, condensedG in zip(binary_Gs, condensed_Gs):
        # binary checks
        assert np.array_equal(np.unique(nx.to_numpy_array(binaryG)), [0, 1])
        assert np.array_equal(np.unique(nx.to_numpy_array(condensedG)), [0, 1])
```
You could easily split this into two loops also.   Other places with `index` can use `zip` as well.
More readable, should be faster because done in C, more "pythonic" (whatever that means).

MaxBromberg replied:
I implemented the suggested changes, though I didn't find other places which could better use `zip`; I think you might've caught them all already.

MaxBromberg commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1166 @@
+                else [0] * column_sums.size
+                for index in range(column_sums.size)
+            ]
+        ).T
```
Clever! I fear I haped upon this convoluted mean myself, because I ran into some error originally, likely solved by your `keepdims=True`. Implemented as suggested, with axis replacing row_normalize and sklearn comment deleted.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1139 @@
+    # Could be passed in most contexts to reduce redundant computation
+    L_GC = len(recursive_leaf_removal(DAG))
+
+    if forward_entropy:
+        B_prime = _matrix_normalize(nx.to_numpy_array(dag), axis=1)
+        # +1 as k \in ( 1, L(G_C) )
+        P = sum([np.power(B_prime, k) for k in range(1, L_GC + 1)])
+        # TODO: Not so sure about this unless k coincides with the number of steps already taken (and the sum is odd)
+        # (Presently awaiting response from original authors for clarification)
+    else:
+        # as w sklearn: B = normalize(nx.to_numpy_array(dag), axis=0, norm='max')  # Column normalization
+        B = _matrix_normalize(nx.to_numpy_array(dag), axis=0)
+        P = sum([np.power(B, k) for k in range(1, L_GC + 1)])
```
I think this whole section collapses to a couple of lines now that axis is the only thing that changes.
The comments indicate that we maybe should be using `max` instead of `sum` to normalize. 
(i.e. use the infinity norm instead of the L_1 norm).   Thoughts?

If this does collapse, we should probably inline the _matrix_normalize

MaxBromberg replied:
Yes! I've collapsed this as suggested. The sklearn `max` norm would've been an improper implementation, as **B** is defined as the adjacency matrix normalized by total _in_ (backward entropy) or _out_ (forward entropy) degree. 
It appears `hierarchy_coordinates.py` does not use `_matrix_normalize` more than the once, though I'm not quite sure how to inline it. (Would seem to have too much state for a lambda function)

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1139 @@
+
+    >>> num_thresholds = 2
+    >>> condensed_networks, _ = nx.node_weighted_condense(b, num_thresholds=num_thresholds)
+    >>> for network_index in range(num_thresholds):
+    ...     for node_index in range(len(condensed_networks[network_index].nodes)):
+    ...         print(f"Node {node_index}, new weight:", condensed_networks[network_index].nodes[node_index]["weight"])
```
```suggestion
    >>> for cG in condensed_networks:
    ...     for node_index, (node, weight) in enumerate(cG.nodes.data("weight")):
    ...         print(f"Node {node}, new weight: {weight}")
```
I couldn't easily see whether the node_index is needed here. If condensed_networks has integer nodes then it isn't.

MaxBromberg replied:
You're right; no need for the `node_index`; implemented as suggested.

MaxBromberg replied:
You're right; no need for the `node_index`; implemented as suggested.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1139 @@
+        )
+
+    successively_peeled_nx_graphs = recursive_leaf_removal(DAG, from_top=False)
+    if (
+        len(successively_peeled_nx_graphs) == 1
+        and len(successively_peeled_nx_graphs[0].nodes()) == 1
+    ):
+        return 0
```
Does this special case treatment avoid an error later? If not, this could be removed...
If so, then don't we need to check all peeled graphs for a single node?

MaxBromberg replied:
So long as `recursive_leaf_removal` is passed DAGs (with `keep_linkless_layer=False` implicitly), the only time a singleton graph will be passed is when the original DAG itself was a single node, and thus the trivial 0 feedforwardness. I think this occurs not infrequently for well connected graphs whose giant component condenses to a single node, and so this 0 feedforwardness weighting is needed in the final average.

MaxBromberg replied:
So long as `recursive_leaf_removal` is passed DAGs (with `keep_linkless_layer=False` implicitly), the only time a singleton graph will be passed is when the original DAG itself was a single node, and thus the trivial 0 feedforwardness. I think this occurs not infrequently for well connected graphs whose giant component condenses to a single node, and so this 0 feedforwardness weighting is needed in the final average.

dschult replied:
Then maybe reduce this (and maybe put before the recursive_leaf_removal) to:   ???
```python
if len(DAG) == 1:
    return 0
```

MaxBromberg replied:
done!

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1139 @@
+
+@not_implemented_for("undirected")
+def feedforwardness(DAG):
+    """Returns feedforwardness hierarchy coordinates, ~extent information passes vertically
+    (vs horizontally) within a hierarchy.
```
The one-line description is used to labels documentation links to this function.
So it really should be short and sweet (one line:)

Here, you should split this an start a next paragraph with the description of what feedforwardness is.
Then  the paragrph you've got about how it is computed -- which should be clarified... it is too vague
as currently written.  What are the weights, how is the "repeated calculation" combined to get a single value (or is it combined?)

MaxBromberg replied:
I wrote up something more of a description, though perhaps the added math is a bit much. Let me know if the description could be improved.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1139 @@
+    float:
+        feedforwardness, hierarchy coordinate
+
+    Examples
```
Should this example have some text describing what is going on?
- The matrix is hard to understand as a graph -- perhaps create a network graph by adding weighted edges?
- What are the 8 graphs that are produced and how are they related to the original?

Maybe there should be a big picture paragraph somewhere in this docstring to explain what this function does to a graph.

MaxBromberg replied:
I've simplified the example to just the infographic, without weighed edges; perhaps that better conveys the essence of it. 
The 8 graphs from the old example result from the default 8 (linear) thresholds produced from `node_weighted_condense`, a lá
```
nx_graphs = [nx.from_numpy_matrix(np.where(A > threshold, 1, 0), create_using=nx.DiGraph) for threshold in thresholds]
```
 I'd add some more big picture explanation, though I'm not sure where's appropriate.

dschult replied:
Add big picture description everywhere you can. It's really needed in this module. Users come to the docs via google searches so you can't tell where they will start. You need to make every doc_string be good enough to be the first doc_string people see.  If you end up repeating ideas in many functions, use the same language in each so the user recognizes that it is the same ideas as they saw before -- but if its the first time they understand what's going on.
:}

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1133 @@
+    L_GC = len(recursive_leaf_removal(DAG))
+
+    axis = 1 if forward_entropy else 0
+    B = _matrix_normalize(nx.to_numpy_array(dag), axis=axis)
```
```suggestion
    B = nx.to_numpy_array(dag)
    # normalize so rows (forward_entropy) or columns (backward_entropy) sum to 1. 
    sums = B.sum(axis=axis, keepdims=True)
    sums[sums==0] = 1
    B = B / sums
```
This is all I was thinking of for "inlining"...

MaxBromberg replied:
done!

MaxBromberg commented:
At present, neither [Prof. Sole](http://complex.upf.es/~ricard/Main/RicardSole.html) nor [Dr. Rodriguez](https://www.linkedin.com/in/carlos-rodriguez-caso-836a3514) have responded to my email asking for clarification on the entropy calculation for this original algorithm, though otherwise I imagine we're roughly ready for release. Is there something I ought to do to better prepare the code at present?

Relevant Emails I found:
ricard.sole@upf.edu
carlos.rodriguez@upf.edu

dschult commented:
I don't think we should rely on a response from the authors if we haven't got one yet.
What is the issue/question a bout entropy that you are asking them about?

MaxBromberg commented:
This was the body of the question I sent them:

> The LaTeX(t) suggests a entropy path is a successive sum for every non-maximal node in a given path over P_ij, the probability of a walker on node i going to node j, multiplied by log(degree of non-maximal node). As I understand the text however, this would necessitate P to be calculated independently for each layer, and not as a sum, i.e: eq[13] P(v_j <-- v_i) = \sum_{1 \leq k \leq L(G_C)} ( [B^T]^k (G) )_ij would then instead become P(v_j <-- v_i) = ( [B^T]^k (G) )_ij,  k = num steps already taken. 
> The infographic's above sums appear to suggest a path's entropy is 1/(num divergences) * log (num divergences), but the sums above do not add to the values showed below unless log = log base 2, not ln; is this an arbitrary choice?

I imagine that the choice of log base is relatively arbitrary, but the difference between the infographic entropy, as implemented, and the different versions of the probability of transition (i.e. with or without the sum over the exponentiated edge normalized adjacency matrix, B) is a difference between a computationally easy analytic calculation (matrix exponentiation) or a sum over all (simple) paths.

dschult commented:
That is helpful, but I still have questions. First of all, does the choice of logarithmic base cause the difference in formulas?  My intuition is that the formula should be the same either way, just with a different base.  The change is base is simply a difference of multiplying by a constant: log_2(x) = ln(x)/ln(2). and my intuition of entropy definitions is that you can multiply by a constant and the important features don't change.  Is that correct?

Can you describe again what drives the difference in formulas?  I don't understand why P would need to be calculated independently for each layer.  Or rather it seems to me that P_ij is computed for each edge. Then the entropy path is computed by walking the path and getting a weighted sum of the edges weighted by log(degree of outgoing node).

MaxBromberg commented:
> That is helpful, but I still have questions. First of all, does the choice of logarithmic base cause the difference in formulas? My intuition is that the formula should be the same either way, just with a different base. The change is base is simply a difference of multiplying by a constant: log_2(x) = ln(x)/ln(2). and my intuition of entropy definitions is that you can multiply by a constant and the important features don't change. Is that correct?

Yes! The choice of log is naught to worry about, methinks. Fundamentally, the character of entropy remains intact.

> Can you describe again what drives the difference in formulas? I don't understand why P would need to be calculated independently for each layer. Or rather it seems to me that P_ij is computed for each edge. Then the entropy path is computed by walking the path and getting a weighted sum of the edges weighted [multiplied] by log(degree of outgoing node).

This is the idea in both cases; entropy is calculated by walking every path between extremal nodes, while maintaining a running sum of the probability of stepping to a given (j th) node multiplied by its degree. 
In `infographic_graph_entropy`, this is done directly, considering every path explicitly thus the `nx.all_simple_paths` call, where at every branch the (integer) number of possible directions is added to a running sum, let's call it **F** (for forks), which then factors into the sum over each path as log(**F**) / **F**.  This sum is then divided by the starting number of extremal nodes  to yield the entropy. Notice that the probability of transition is here only _implicitly_ considered, as entropy is a literal weighted sum of possible forks.

The `graph_entropy` call is hopefully a bit more clever, and computationally concise. It uses the fact that the n-th power of a degree-normalized adjacency matrix yields a matrix whose i,j-th element is the probability of transition from node i to node j after n steps. This means the sum of the powers of the degree normalized matrix (` P = sum([np.power(B, k) for k in range(1, L_GC + 1)]) `)  yield the probability of all possible transitions, accounting for more or less direct paths, with proportionately varying number of steps, between any two given nodes. 
Therefore, the sum over this probability matrix **P** need only consider each intermediary node once per maximal, minimal node pair, and all possible paths need not be calculated explicitly, as via:
```
e = sum(P[u][v] * np.log(deg(u)) for u in boundary_layer for v in non_extremal_nodes)
```
The difference is that `infographic_graph_entropy` (with present implementation) calculates single layer entropy, equivalent to the `graph_entropy` algorithm using only extremal nodes and powers of **B** independently (not summed). By using the summed powers of **B**, (**P**) in `graph_entropy` , and evaluating over all nodes (not just extremal nodes, but all paired with starting nodes) we consider all pruned layers as well. That said, I'm not sure which is intended; treeness does consider all layers independently in the infographic, but seemingly collectively in computation. (specifically in the [appendix][1], eq. 15 takes into account all possible transition paths, which in eq. 16, appears to be evaluated per node)

This said, I just saw I need to change a couple things in the `graph_entropy` function, specifically replacing u with v in `np.log(deg(u))` above, and ensuring that `B.T` is exponentiated for backwards entropy. I'll push these changes soon, but it doesn't change the main point of confusion. Despite all this, both forms of graph entropy make sense to me, and it seems a relatively minor semantic difference in the grand scheme of getting a clear notion of hierarchy coordinates.

[1]: www.pnas.org/lookup/suppl/doi:10.1073/pnas.1300832110/-/DCSupplemental

dschult commented:
Glad that this is getting closer to matching the mathematical description in the paper.

Not that you need more to think about here, but I'm concerned with network algorithms that rely on formulas that use matrix exponentials.  The formulas themselves are usually compact and easy to read (yay for linear algebra!). But in a network setting the matrix is usually sparse -- and the equivalent of exponentiation can be done effectively by traversing the graph rather than forming a (dense) matrix and exponentiating it.  

Are these methods intended to be used on fairly dense networks? What is a typical use case for this algorithm?

Also, am I missing something that makes the linear algebra approach stronger here even when we don't have dense networks/matrices?

MaxBromberg commented:
> Not that you need more to think about here, but I'm concerned with network algorithms that rely on formulas that use matrix exponentials. The formulas themselves are usually compact and easy to read (yay for linear algebra!). But in a network setting the matrix is usually sparse -- and the equivalent of exponentiation can be done effectively by traversing the graph rather than forming a (dense) matrix and exponentiating it.

Good point! This is something which I addressed in an earlier use of matrix exponentiation for Random Walker Effective Distances using scipy's sparse optimization (and could be likewise adapted into this implementation when matrix exponentiation is called for) but I was under the mistaken impression that matrix exponentiation is easier computationally than repeated `all_simple_paths` calls. 

> Are these methods intended to be used on fairly dense networks? What is a typical use case for this algorithm?

There is no particular reason why hierarchy coordinates would be used more in dense networks.

> Also, am I missing something that makes the linear algebra approach stronger here even when we don't have dense networks/matrices?

...Likely not, and I'm mistaken as to which is more efficient. However, just rephrasing my earlier note, my reasoning was that the (sum of) matrix exponentiation(s) occurs but once, and then for each pair of extremal nodes, each node is considered but once. In the `nx.all_simple_paths` dependent case (`infographic_graph_entropy`) , for each pair of extremal nodes there's a new `nx.all_simple_paths` call, and all paths are evaluated individually, such that paths which are identical save for a single deviation nonetheless are computed wholly independently. 
From what a quick reference to [Wikipedia][1] suggests, square matrix multiplication of dims (n x n)^k would take at best ~O(k * n^2.4) time, whereas the `nx.all_simple_paths` algorithm (constant, O(V+E) time) would be called |M| * |mu| times, where |M| and |mu| are the number of maximal, minimal nodes respectively. ...Which just goes to show why I should've considered this earlier; it appears it would be an extraordinary circumstance indeed which leads to a faster matrix computation than simply using the existing `infographic_graph_entropy` call. 
My bad then! Apart from the possible change of log base, increasing tests, relabeling `infographic_graph_entropy` to `graph_entropy`,  `graph_entropy` to `analytic_graph_entropy`, and dealing with whatever made my last push fail to pass all checks, we might be good. Let me know if that seems a reasonable conclusion, or what else it might need.

[1]: https://www.wikiwand.com/en/Computational_complexity_of_mathematical_operations

MaxBromberg commented:
Just

MaxBromberg commented:
Just re-iterativing this, I might've slipped by in my edit rather than repost:
> Apart from the possible change of log base, increasing tests, relabeling infographic_graph_entropy to graph_entropy, graph_entropy to analytic_graph_entropy, and dealing with whatever made my last push fail to pass all checks, we might be good. Let me know if that seems a reasonable conclusion, or what else it might need.

dschult commented:
My head has not been able to reload this thread completely yet...
But your last comment does indeed sound reasonable.  

I think the last two tests are not passing due to a fix that has been merged with the main branch. To get them to pass probably just requires merging with the main branch again.

MaxBromberg commented:
Thanks for the step by step explanation; I hope you have a copy pasta for the confused like myself. 
`git remote -v` shows the upstream already as `git@github.com:networkx/networkx.git`, and `git pull upstream main` results in an access error:
```
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
Though I thought read access was allowed...

dschult commented:
That is indeed confusing...  Can you check that connecting to other similar repositories still works correctly? I know [github is making changes to its authentication rules,](https://github.blog/2020-07-30-token-authentication-requirements-for-api-and-git-operations/) but this doesn't seem like it is related. And this worked before, so that's confusing.

I guess you could try making a new remote using a different interface (like https://github.com/networkx/networkx.git)-- or even deleting and re-adding the remote git@github.com.  The repository definitely should give read access for everyone.

MaxBromberg commented:
Alright! Hopefully this works, I took your advice and used  `https://github.com/networkx/networkx.git` rather than `git@github.com...` and it seemed to then allow the `git pull upstream main`.

dschult commented on networkx/algorithms/tests/test_hierarchy_coordinates.py:
```diff
@@ -0,0 +1,273 @@
+
+
+def test_hierarchy_coordinate():
+    import numpy as np
```
Shouldn't need import numpy since imported at top.

dschult commented on networkx/algorithms/tests/test_hierarchy_coordinates.py:
```diff
@@ -0,0 +1,273 @@
+    assert b_hc == [-0.56, 0.56, 0.43]
+
+
+test_hierarchy_coordinate()
```
Let pytest run the test (it also loads the local repository).

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1107 @@
+========================
+Functions to calculate hierarchy coordinates for directed networks.
+
+Hierarchy coordinates consist of treeness, feedforwardness, and orderability, which collectively create a
```
Can you pull the docs into an 80-char width (or so) for better readability?

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1107 @@
+
+# Hierarchy Coordinate Functions
+def node_weighted_condense(A, num_thresholds=8, threshold_distribution=None):
+    """Creates a series of networkx graphs based on generated edge weight thresholds, and returns their node weighted
```
Can you make the doc_strings start with a single-line description. This is used as text to go with links to this function in the documentation.  Keep it short.

Then follow up with a paragraph that actually says more completely what the function does.  Try to keep all doc lines around 80-chars or less.

You need to define what "node weighted condense and original graphs" means.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1107 @@
+
+    Note:
+    ------
+    TODO: Might wish to eliminate return, or enable copying?
```
If it is changed "in-place" then don't return the graph. Also state in the documentation that the graph is updated in place.

MaxBromberg replied:
I wasn't sure how to best implement this; presumably as a member of the graph class, such that a G.weight_nodes_by_condensation would make sense, but I think it might be better to either do without the function (it's only used once), or make it general enough to be generally useful. Or it could be best as an optional argument in the condense function?

MaxBromberg replied:
I wasn't sure how to best implement this; presumably as a member of the graph class, such that a G.weight_nodes_by_condensation would make sense, but I think it might be better to either do without the function (it's only used once), or make it general enough to be generally useful. Or it could be best as an optional argument in the condense function?

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1107 @@
+def max_min_layers(G, max_layer=True):
+    """Returns the maximal (k_in = 0) layer or the minimal layer (k_out = 0)
+
+    Returns the maximal (k_in = 0, highest in hierarchy) layer (those nodes with in degree = 0)
+     or the minimal layer (k_out = 0) of a directed network.
```
Maybe say what the words hierarchy and layer mean?  Just give some context.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1107 @@
+    NetworkXError
+        If input not a directed graph
+
+    Notes:
+    -------
+
```
If you don't have notes, you can leave that section out.

dschult commented on networkx/algorithms/hierarchy_coordinates.py:
```diff
@@ -0,0 +1,1107 @@
+
+@not_implemented_for("undirected")
+def leaf_removal(G, top=True):
+    """Returns a pruned network, with either maximal (k_in=0)
```
Tighten up this language.  And add a paragraph to describe it more fully. What do you mean by pruned? Say something about why this function is useful, or how it fits wit the other code around it -- when should users use it?

MaxBromberg commented:
Thanks for the suggestions, which I hope I've made appropriate edits for, save re: weight_nodes_by_condensation, as commented above. I hope this makes hierarchy coordinates nearly deployable, at least as the original author's used them. 

There remains a significant edit to accommodate the edge cases addressed in a paragraph of their appendix (p. 7) when they consider how condensation can break the graph into more than one single connected component. In their numerical investigations, they only consider the largest connect components, however theoretically outline taking a weighted average of all components, which we should perhaps likewise do. In code, this would translate to accommodating the consequences of changing
```
    for condensed_graph in condensed_graphs:
        largest_condensed_graphs.append(nx.convert_node_labels_to_integers(condensed_graph.subgraph(
                    max(nx.weakly_connected_components(condensed_graph), key=len))).copy())
```
to 
```
    for condensed_graph in condensed_graphs:
        condensed_graphs_per_threshold.append([nx.convert_node_labels_to_integers(condensed_graph.subgraph(c)) 
for c in sorted(
        nx.weakly_connected_components(condensed_graph), key=len, reverse=True).copy()])
                                              )
```

MaxBromberg commented:
Thanks for the suggestions, which I hope I've made appropriate edits for, save re: weight_nodes_by_condensation, as commented above. I hope this makes hierarchy coordinates nearly deployable, at least as the original author's used them. 

There remains a significant edit to accommodate the edge cases addressed in a paragraph of their appendix (p. 7) when they consider how condensation can break the graph into more than one single connected component. In their numerical investigations, they only consider the largest connect components, however theoretically outline taking a weighted average of all components, which we should perhaps likewise do. In code, this would translate to accommodating the consequences of changing
```
    for condensed_graph in condensed_graphs:
        largest_condensed_graphs.append(nx.convert_node_labels_to_integers(condensed_graph.subgraph(
                    max(nx.weakly_connected_components(condensed_graph), key=len))).copy())
```
to 
```
    for condensed_graph in condensed_graphs:
        condensed_graphs_per_threshold.append([nx.convert_node_labels_to_integers(condensed_graph.subgraph(c)) 
for c in sorted(
        nx.weakly_connected_components(condensed_graph), key=len, reverse=True).copy()])
                                              )
```

dschult commented:
I think you are right that the proper treatment is averaging over components. 
I think your code example has a `copy()` misplaced -- it should copy the subgraph, not the sorted list. Otherwise it looks good to me.

I think there are a couple of places in the newly revised docstrings where the indent level changes. Can you correct those? And also a place with 3 blank lines before the "See Also" section.  Perhaps reduce to 1 or 2?

Thanks for this!

MaxBromberg commented:
I'm sorry I haven't returned to this in very long! I've been busy with a new job, and have some difficulty reloading the project into mental RAM, but I promise I'm alive and aware I ought to!

dschult commented:
No problem Maqz!  (Darn that real life stuff -- it keeps getting in the way!)

Quick question in case you know the answer easily.  The docs say that hierarchy coordinates can classify all networks structures. "all" is a strong word. Does this mean that the three hierarchy sets of values uniquely identify the network?
I'm just try to find the right words there...
Thanks!
